### PR TITLE
[GEP-26] Garden Security Accessor and ProviderType predicate for CredentialsBinding

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**How to categorize this PR?**

/area dev-productivity quality ipcei
/kind enhancement
/label ipcei/workload-identity

**What this PR does / why we need it**:
A new predicate `extensions/pkg/predicate.GardenSecurityProviderType` can be used to select resources from the `security.gardener.cloud` group that are related to the passed provider type.

This is useful in provider extension admission webhooks to filter related resources. See https://github.com/gardener/gardener-extension-provider-aws/blob/5a0ca81b53918509d147b0e0c141480fdc4266a3/pkg/admission/validator/webhook.go#L37

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/9586

Related to https://github.com/gardener/gardener-extension-provider-aws/pull/1047

**Special notes for your reviewer**:

cc @vpnachev 

ping @timuthy because we had an offline discussion in case you want to take a look.

I am not sure if versioned APIs need to implement these interfaces. I noticed that `v1beta1` Cloudprofiles do not implement the `gardencore.Object` interface, but `v1beta1` Shoots do implement the interface.
See https://github.com/gardener/gardener/blob/46e8315c60548020cf6c01c247dc143e7009a04d/pkg/apis/core/v1beta1/types_cloudprofile.go#L18 and https://github.com/gardener/gardener/blob/46e8315c60548020cf6c01c247dc143e7009a04d/pkg/apis/core/v1beta1/types_shoot.go#L144.

**Release note**:

```noteworthy developer
A new predicate `extensions/pkg/predicate.GardenSecurityProviderType` can be used to select resources from the `security.gardener.cloud` group that are related to the passed provider type.
```

### CODE REVIEW

File: pkg/api/security/accessor.go [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// Accessor tries to create a `gardensecurity.Object` from the given runtime.Object.
+//
+// If the given object already implements Object, it is returned as-is.
+// Otherwise, an error with the type of the object is returned.
+func Accessor(obj runtime.Object) (gardensecurity.Object, error) {
+	switch v := obj.(type) {
+	case gardensecurity.Object:
+		return v, nil
+	default:
+		return nil, fmt.Errorf("value of type %T does not implement Object", obj)
```
Comment (vpnachev) [inline pkg/api/security/accessor.go]: Maybe the error message will be more helpful it if precise the originating package of `Object`
```suggestion
		return nil, fmt.Errorf("value of type %T does not implement gardensecurity.Object", obj)
```

**Review by vpnachev — COMMENTED** [review]
Looks good to me, just one in-line suggestion.

File: pkg/api/security/accessor.go (hunk 1 above)
Comment (dimityrmirchev) [reply pkg/api/security/accessor.go]: Nice suggestion! Done with https://github.com/gardener/gardener/pull/10499/commits/df45bafa277b72d7e61cc8ba0b23d51de275c6c2

**Review by plkokanov — COMMENTED** [review]
One minor question/nit, otherwise lgtm.
Shouldn't this file: https://github.com/gardener/gardener/blob/df45bafa277b72d7e61cc8ba0b23d51de275c6c2/pkg/api/security/accessor_suite_test.go be named `security_suite_test.go`?

**Comment by dimityrmirchev** [comment]
> One minor question/nit, otherwise lgtm. Shouldn't this file: https://github.com/gardener/gardener/blob/df45bafa277b72d7e61cc8ba0b23d51de275c6c2/pkg/api/security/accessor_suite_test.go be named `security_suite_test.go`?

Yes, I will rename it

**Pull request merged**